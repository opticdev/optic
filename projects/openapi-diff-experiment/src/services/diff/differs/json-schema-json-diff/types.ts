import { BodyLocation, OpenAPIV3 } from '@useoptic/openapi-utilities';
import Ajv, { ErrorObject } from 'ajv';
import { ShapeDiffTypes } from '../../types';
import { JsonSchemaPatch } from './plugins/plugin-types';
import { JsonPatcher } from '../../../patch/incremental-json-patch/json-patcher';

export interface JsonSchemaJsonDiffer {
  ajvToDiff: (
    schema: OpenAPIV3.SchemaObject,
    schemaPath: string,
    diff: ErrorObject,
    example: any,
    location: BodyLocation
  ) => ShapeDiffTypes;
  diffToPatch: (
    diff: ShapeDiffTypes,
    patcher: JsonPatcher<OpenAPIV3.Document>
  ) => JsonSchemaPatch[];
  compare: (
    schema: OpenAPIV3.SchemaObject,
    to: any,
    location: BodyLocation,
    schemaPath: string,
    options: {
      collapseToFirstInstanceOfArrayDiffs: boolean;
    }
  ) => ShapeDiffTypes[];
  supportedKeywords: string[];
  opticAjv: Ajv;
  stripKeywordsNotSupported: (diffs: ErrorObject[]) => ErrorObject[];
}
