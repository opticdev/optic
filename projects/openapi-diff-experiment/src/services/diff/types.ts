/*

  Interface diffing traffic
   - produces a diff from the
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
import { ApiTraffic } from '../traffic/types';
import { OpenAPIDiffingQuestions } from '../read/types';
import {
  ConceptualLocation,
  FieldLocation,
  OpenAPIV3,
} from '@useoptic/openapi-utilities';
import { JsonSchemaKnownKeyword } from './differs/json-schema-json-diff/plugins/plugin-types';
import { JsonSchemaJsonDiffer } from './differs/json-schema-json-diff/types';
import { JsonPath } from '@useoptic/openapi-io';

export interface IDiffService {
  compare(traffic: ApiTraffic): Promise<{ diffs: IDiff[]; errors: string[] }>;
  jsonSchemaDiffer: JsonSchemaJsonDiffer;
}

export type DiffServiceFactoryFunction = (
  spec: OpenAPIDiffingQuestions
) => IDiffService;

export enum DiffType {
  UnmatchedPath = 'UnmatchedPath',
  UnmatchedMethod = 'UnmatchedMethod',
  UnmatchedResponse = 'UnmatchedResponse',
  BodyAdditionalProperty = 'BodyAdditionalProperty',
  BodyMissingRequiredProperty = 'BodyMissingRequiredProperty',
  BodyUnmatchedType = 'BodyUnmatchedType',
  QueryAdditionalParameter = 'QueryAdditionalParameter',
}

export interface IDiff {
  type: DiffType;
}

export interface UnmatchedPath extends IDiff {
  type: DiffType.UnmatchedPath;
  path: string;
  method: OpenAPIV3.HttpMethods;
  closestMatch: string;
}

export interface UnmatchedMethod extends IDiff {
  type: DiffType.UnmatchedMethod;
  path: string;
  method: OpenAPIV3.HttpMethods;
}

export interface UnmatchedResponse extends IDiff {
  type: DiffType.UnmatchedResponse;
  path: string;
  method: OpenAPIV3.HttpMethods;
  statusCode: string;
}
////////////////////////////////////////////////////////////

export interface QueryAdditionalParameter extends IDiff {
  type: DiffType.QueryAdditionalParameter;
  path: string;
  name: string;
  method: OpenAPIV3.HttpMethods;
  example: any;
}

////////////////////////////////////////////////////////////

export interface BodyAdditionalProperty extends IDiff {
  type: DiffType.BodyAdditionalProperty;
  propertyExamplePath: JsonPath;
  parentObjectPath: JsonPath;
  key: string;
  example: any;
  propertyPath: JsonPath;
  keyword: JsonSchemaKnownKeyword.additionalProperties;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export interface BodyMissingRequiredProperty extends IDiff {
  type: DiffType.BodyMissingRequiredProperty;
  keyword: JsonSchemaKnownKeyword.required;
  parentObjectPath: JsonPath;
  propertyPath: JsonPath;
  key: string;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export interface BodyPropertyUnmatchedType extends IDiff {
  type: DiffType.BodyUnmatchedType;
  keyword: JsonSchemaKnownKeyword.type | JsonSchemaKnownKeyword.oneOf;
  propertyPath: JsonPath;
  key: string;
  example: any;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export type ShapeDiffTypes =
  | BodyMissingRequiredProperty
  | BodyAdditionalProperty
  | BodyPropertyUnmatchedType;

type DiffKinds =
  | UnmatchedMethod
  | UnmatchedPath
  | UnmatchedResponse
  | QueryAdditionalParameter
  | ShapeDiffTypes;

/*
  Diff Result Types
 */

export interface DiffResult {
  isMatch: boolean;
}
export interface MatchedWithContext<Context> extends DiffResult {
  isMatch: true;
  context: Context;
}
export interface Matched extends DiffResult {
  isMatch: true;
}

export interface Diff extends DiffResult {
  isMatch: false;
  diffs: DiffKinds[];
}

export interface DiffError extends DiffResult {
  isMatch: false;
  error: string;
}

export interface EitherDiffResult<Context = undefined> {
  context: Context;
  error: string;
  diffs: DiffKinds[];
  isMatch: boolean;
  isDiff: boolean;
}

export function EitherDiffResult<Context>(
  i: DiffResult
): EitherDiffResult<Context> {
  return {
    isDiff: !i.isMatch,
    isMatch: i.isMatch,
    context: i.hasOwnProperty('context')
      ? (i as MatchedWithContext<Context>).context
      : undefined,
    error: i.hasOwnProperty('error') ? (i as DiffError).error : undefined,
    diffs: i.hasOwnProperty('diffs') ? (i as Diff).diffs : [],
  };
}

export const DiffResult = {
  match: () => {
    return EitherDiffResult<undefined>({ isMatch: true } as Matched);
  },
  matchWithContext: <Context>(context: Context) => {
    return EitherDiffResult<Context>({
      isMatch: true,
      context,
    } as MatchedWithContext<Context>);
  },
  diff: (diffs: DiffKinds[]) => {
    return EitherDiffResult<undefined>({
      isMatch: false,
      diffs,
    } as Diff);
  },
  error: (error: string) => {
    return EitherDiffResult<undefined>({ isMatch: false, error } as DiffError);
  },
};

export function isSchemaDiff(i: IDiff): boolean {
  return (
    i.type === DiffType.BodyAdditionalProperty ||
    i.type === DiffType.BodyUnmatchedType ||
    i.type === DiffType.BodyMissingRequiredProperty
  );
}
