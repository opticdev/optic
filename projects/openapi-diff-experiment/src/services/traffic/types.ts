import events from 'events';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
/*

  Interface for traffic, expose what the differ cares about via interfaces
   - different formats can have different implementations
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
export interface ApiTraffic {
  method: OpenAPIV3.HttpMethods;
  path: string;
  queryString: string;
  response: {
    statusCode: string;
    body: {
      contentType?: string;
      jsonBodyString?: string;
    };
  };
  requestBody?: {
    contentType?: string;
    jsonBodyString?: string;
  };
}

/*

  Interface for traffic, should get traffic (somehow), and emit each interaction into our queue
   - each implementation should be self contained starting / parsing / stopping traffic collection
   - known types
     - file
     - in-memory (for debugging)
     - proxy
     - postman
     - sniff
     - chrome

   */

export interface ITrafficSource {
  start(): Promise<void>;
  stop(): Promise<void>;
  emitTraffic(traffic: ApiTraffic): Promise<void>;
  on(event: 'traffic', listener: (example: ApiTraffic) => void): this;
}

export class TrafficSource
  extends events.EventEmitter
  implements ITrafficSource
{
  emitTraffic(traffic: ApiTraffic): Promise<void> {
    this.emit('traffic', traffic);
    return Promise.resolve();
  }

  start(): Promise<void> {
    return Promise.resolve(undefined);
  }

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}
