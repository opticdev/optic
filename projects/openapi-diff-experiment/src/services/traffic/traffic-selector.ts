import { ApiTraffic } from './types';
import { ConceptualLocation } from '@useoptic/openapi-utilities';
import invariant from 'ts-invariant';

export function trafficSelector(traffic: ApiTraffic) {
  return {
    bodyJsonForLocation: (location: ConceptualLocation) => {
      if ('inResponse' in location && 'body' in location.inResponse) {
        const contentType = location.inResponse.body.contentType;

        invariant(
          contentType === traffic.response.body.contentType,
          'content type must match when selecting examples from traffic'
        );

        invariant(
          typeof traffic.response.body.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );

        return JSON.parse(traffic.response.body.jsonBodyString!);
      }

      if ('inRequest' in location && 'body' in location.inRequest) {
        const { contentType } = location.inRequest.body;
        invariant(
          contentType === traffic.requestBody?.contentType,
          'content type must match when selecting examples from traffic'
        );
        invariant(
          typeof traffic.requestBody.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );
        return JSON.parse(traffic.requestBody.jsonBodyString!);
      }

      invariant(false, 'can not select example body from traffic');
    },
  };
}
