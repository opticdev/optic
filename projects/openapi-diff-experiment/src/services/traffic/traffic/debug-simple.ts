import { ApiTraffic } from '../types';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';

export class SimpleExampleTraffic implements ApiTraffic {
  constructor(
    private _request: { method: string; path: string; queryString?: string },
    private _response: { statusCode: string; json?: any }
  ) {}
  get method() {
    return this._request.method.toLowerCase() as OpenAPIV3.HttpMethods;
  }
  get path() {
    return this._request.path;
  }
  get response() {
    return {
      statusCode: this._response.statusCode,
      body: {
        contentType: this._response.json ? 'application/json' : undefined,
        jsonBodyString: this._response.json
          ? JSON.stringify(this._response.json)
          : undefined,
      },
    };
  }

  get queryString() {
    return this._request.queryString || '';
  }
}

export function makeExample(
  path: string,
  method: string | OpenAPIV3.HttpMethods,
  statusCode: string = '200',
  jsonBody?: object | undefined
) {
  return new SimpleExampleTraffic(
    { path, method },
    { statusCode, json: jsonBody }
  );
}
