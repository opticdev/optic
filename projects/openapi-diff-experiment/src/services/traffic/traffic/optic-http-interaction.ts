import { ApiTraffic } from '../types';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';

// TODO: Fill in the type
// {
//   "uuid": "c11116b1-cea5-401c-b151-da47ab43cee3",
//   "request": {
//     "host": "httpbin.org",
//     "method": "GET",
//     "path": "/get",
//     "headers": {
//       "shapeHashV1Base64": null,
//       "asJsonString": null,
//       "asText": null
//     },
//     "query": {
//       "shapeHashV1Base64": null,
//       "asJsonString": null,
//       "asText": "a=1"
//     },
//     "body": {
//       "contentType": null,
//       "value": {
//         "shapeHashV1Base64": null,
//         "asJsonString": null,
//         "asText": null
//       }
//     }
//   },
//   "response": {
//     "statusCode": 200,
//     "headers": {
//       "shapeHashV1Base64": null,
//       "asJsonString": null,
//       "asText": null
//     },
//     "body": {
//       "contentType": "application/json",
//       "value": {
//         "shapeHashV1Base64": "CAASEwoEYXJncxILCAASBwoBYRICCAISUAoHaGVhZGVycxJFCAASDAoGQWNjZXB0EgIIAhIKCgRIb3N0EgIIAhIQCgpVc2VyLUFnZW50EgIIAhIVCg9YLUFtem4tVHJhY2UtSWQSAggCEgwKBm9yaWdpbhICCAISCQoDdXJsEgIIAg==",
//         "asJsonString": "{\"args\":{\"a\":\"1\"},\"headers\":{\"Accept\":\"*/*\",\"Host\":\"httpbin.org\",\"User-Agent\":\"curl/7.64.1\",\"X-Amzn-Trace-Id\":\"Root=1-618d1bfc-0c9eea1060c29d6348b62f19\"},\"origin\":\"31.10.148.195\",\"url\":\"http://httpbin.org/get?a=1\"}",
//         "asText": null
//       }
//     }
//   },
//   "tags": []
// }
type IOpticCaptureHttpInteraction = any;

export class OpticHttpInteraction implements ApiTraffic {
  private _method: OpenAPIV3.HttpMethods;
  private _path: string;
  private _queryString: string;
  private _requestBody: ApiTraffic['requestBody'];
  private _response: {
    statusCode: string;
    body: {
      contentType: string | undefined;
      jsonBodyString: string | undefined;
    };
  };

  constructor(_interaction: IOpticCaptureHttpInteraction) {
    this._method =
      _interaction.request.method.toLowerCase() as OpenAPIV3.HttpMethods;
    this._path = _interaction.request.path;
    this._queryString = _interaction.request.query.asText || '';
    this._requestBody = _interaction.request.body && _interaction.request.body.contentType
      ? {
          contentType: _interaction.request.body.contentType,
          jsonBodyString:
            _interaction.request.body.value.asJsonString || undefined,
        }
      : undefined;
    this._response = {
      statusCode: _interaction.response.statusCode.toString(),
      body: {
        contentType: _interaction.response.body && _interaction.response.body.contentType
          ? 'application/json'
          : undefined,
        jsonBodyString: _interaction.response.body.value.asJsonString
          ? _interaction.response.body.value.asJsonString
          : undefined,
      },
    };
  }

  get method() {
    return this._method;
  }
  get path() {
    return this._path;
  }
  get requestBody() {
    return this._requestBody;
  }
  get response() {
    return this._response;
  }

  get queryString() {
    return this._queryString;
  }
}
