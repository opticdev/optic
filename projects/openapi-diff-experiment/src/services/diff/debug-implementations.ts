import { IDiff, IDiffService } from './types';
import { ApiTraffic } from '../traffic/types';
import { opticJsonSchemaDiffer } from './differs/json-schema-json-diff';
import { JsonSchemaJsonDiffer } from './differs/json-schema-json-diff/types';

export class NeverReturnDiffService implements IDiffService {
  async compare(
    traffic: ApiTraffic
  ): Promise<{ diffs: IDiff[]; errors: string[] }> {
    return { diffs: [], errors: [] };
  }

  jsonSchemaDiffer: JsonSchemaJsonDiffer = opticJsonSchemaDiffer();
}
