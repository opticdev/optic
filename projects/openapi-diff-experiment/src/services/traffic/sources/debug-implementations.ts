import { ApiTraffic, TrafficSource } from '../types';
import { waitFor } from '../../../utils/debug_waitFor';
import fs from 'fs-extra';
import { OpticHttpInteraction } from '../traffic/optic-http-interaction';

export class DebugSource extends TrafficSource {
  constructor(private examples: ApiTraffic[], private pauseFor: number) {
    super();
  }

  start(): Promise<void> {
    setTimeout(() => {
      this.examples.forEach(async (i) => {
        await this.emitTraffic(i);
        await waitFor(this.pauseFor);
      });
    }, 50);
    return Promise.resolve(undefined);
  }

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}

export class PassThroughSource extends TrafficSource {
  constructor() {
    super();
  }
  start(): Promise<void> {
    return Promise.resolve(undefined);
  }

  sendTraffic = (...traffic: ApiTraffic[]) => {
    traffic.forEach((i) => this.emitTraffic(i));
  };

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}

export class OpticOssDebugCaptureSource extends TrafficSource {
  constructor(private filePath: string, private pauseFor: number) {
    super();
  }

  async start(): Promise<void> {
    const debugCapture = await fs.readJson(this.filePath);

    await waitFor(50);

    for (let interaction of debugCapture) {
      const interactionInstance = new OpticHttpInteraction(interaction);
      await this.emitTraffic(interactionInstance);
      await waitFor(this.pauseFor);
    }

    return Promise.resolve(undefined);
  }

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}
