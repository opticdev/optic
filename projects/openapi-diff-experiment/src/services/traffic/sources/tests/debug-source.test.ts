import { DebugSource } from '../debug-implementations';
import { makeExample } from '../../traffic/debug-simple';
import { ApiTraffic } from '../../types';
import { waitFor } from '../../../../utils/debug_waitFor';

describe('debug sources', () => {
  it('can iterate provided traffic', async () => {
    const source = new DebugSource(
      [
        makeExample('/examples', 'get', '200'),
        makeExample('/examples', 'get', '200'),
        makeExample('/examples/:todoId', 'get', '200'),
        makeExample('/examples', 'post', '404'),
      ],
      10
    );

    const examples: ApiTraffic[] = [];
    source.on('traffic', (example) => {
      examples.push(example);
    });
    await source.start();

    await waitFor(300);

    expect(examples).toMatchSnapshot();
  });
});
