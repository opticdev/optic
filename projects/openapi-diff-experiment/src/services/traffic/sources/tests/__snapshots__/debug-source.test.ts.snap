// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug sources can iterate provided traffic 1`] = `
Array [
  SimpleExampleTraffic {
    "_request": Object {
      "method": "get",
      "path": "/examples",
    },
    "_response": Object {
      "json": undefined,
      "statusCode": "200",
    },
  },
  SimpleExampleTraffic {
    "_request": Object {
      "method": "get",
      "path": "/examples",
    },
    "_response": Object {
      "json": undefined,
      "statusCode": "200",
    },
  },
  SimpleExampleTraffic {
    "_request": Object {
      "method": "get",
      "path": "/examples/:todoId",
    },
    "_response": Object {
      "json": undefined,
      "statusCode": "200",
    },
  },
  SimpleExampleTraffic {
    "_request": Object {
      "method": "post",
      "path": "/examples",
    },
    "_response": Object {
      "json": undefined,
      "statusCode": "404",
    },
  },
]
`;
